### PR TITLE
Fix extract_flow_from_file for flows using emoji

### PR DESCRIPTION
**Pull request by taljaards**
### Summary
Flows can include the use of emojis and be valid (e.g. for creating markdown Artifacts).
Validating that flow using `flow.validate()` works perfectly, and the flow is runnable.

However, when you read that file using `prefect.utilities.storage.extract_flow_from_file`, the file fails to read and a `UnicodeDecodeError` is thrown.

~~This fix isn't the only option to resolve this issue.~~
~~I opted to add a catch and retry if the reading fails. This fully preserves the original behaviour.~~

The fix is to read the file in binary mode. The `exec` a few lines down handles the encoding fine.

### Steps Taken to QA Changes
- Local testing:
  
  Create `sample_flow.py`:
  ```py
  from prefect import task, Flow
  
  @task
  def print_msg():
      print("✅ Success, False: ❌")
  
  with Flow("Sample flow") as flow:
      print_msg()
  
  flow.validate()
  ```
  Create and run `flow_validator.py`:
  ```py
  from prefect.utilities.storage import extract_flow_from_file
  
  flow = extract_flow_from_file("sample_flow.py")
  
  assert flow.name == "Sample flow"
  assert flow.run().is_successful()
  ```

- I added a unit test similar to the above.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- No tests or issue needed
- [x] A short code fix
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a bug report issue 
	- Please include tests. One-line fixes without tests will not be accepted unless it's related to the documentation only.
- [ ] A new feature implementation
	- Please reference the related issue by including "closes `<link to issue>`" in this Pull Request's summary section. 
        - If no issue exists, please create a feature enhancement issue 
	- Please include tests
    - Please make sure that your QA steps are both thorough and easy to reproduce by somebody with limited knowledge of the feature that you are submitting

### CODE REVIEW

**Comment by madkinsz** [comment]
This seems like a pretty reasonable approach. Can we just use `"rb"` though and let `exec` deal with the encoding?

**Comment by taljaards** [comment]
I missed this. Changed in 74bec5d.

**Comment by madkinsz** [comment]
@taljaards looks like I gave a bad suggestion and `exec` will just throw a syntax error? Weird. I think your try/retry approach is reasonable. I think I'd try UTF-8 first then fall back to ASCII. cc @bunchesofdonald and @cicdw for additional opinions on the implementation.

**Comment by taljaards** [comment]
@madkinsz strange indeed. I tested again now -
seems with ❌ in the flow name it gives that syntax error.
Both the `"rb"` method and the try/retry approach works if the ❌ isn't in the flow name, but somewhere inside a task (like my example in the PR description).

I'll keep an eye on here for comments on your suggestion, then make the change accordingly.

**Comment by taljaards** [comment]
@madkinsz I made some changes; see bullets below. I think the current revision is the best.

- Use the `"rb"` method; `exec` now handles it correctly (see bullets below).
- Weird syntax error [fixed](https://github.com/PrefectHQ/prefect/pull/6375/commits/0ca9b32cbd5c72abdaca73fe939a124f3fa22e60).
- Also, with some digging I think I realised why this bug slipped by unnoticed until now - seems Windows default encoding = `cp1252`, and Linux (or WSL where I checked) = `utf8`. So no one picked it up until now. Hence, I also added the encoding option to the save [here](https://github.com/PrefectHQ/prefect/pull/6375/commits/0ca9b32cbd5c72abdaca73fe939a124f3fa22e60).

**Review by madkinsz — APPROVED** [review]
Thank you! This definitely looks ideal.

**Comment by madkinsz** [comment]
Looks like your test file needs a run of `black`. We can figure out these other CI issues... @abrookins I've already asked @desertaxle but we'll need to allocate time to fixing the 1.x CI so I can cut a release on the first Tuesday of September.

**Comment by desertaxle** [comment]
I'm exploring options to fix the 1.x CI in https://github.com/PrefectHQ/prefect/pull/6546

File: src/prefect/utilities/storage.py [hunk 1]
```diff
@@ -79,7 +79,7 @@ def extract_flow_from_file(
             contents = f.read()
 
     elif file_contents is not None:
-        contents = file_contents
+        contents = file_contents.encode("utf-8")
```
Comment (madkinsz) [inline src/prefect/utilities/storage.py:82]: I'm not sure we want to do this. I think we want `exec` to just accept the raw bytes without choosing the encoding. Was there a type-check error or something?

Comment (taljaards) [reply src/prefect/utilities/storage.py:82]: Yes, mypy found: 
```
src/prefect/utilities/storage.py:82: error: Incompatible types in assignment (expression has type "str", variable has type "bytes"
```

Comment (madkinsz) [reply src/prefect/utilities/storage.py:82]: Ah I see here we're then using `bytes` for `contents` consistently. All strings _should_ be utf-8 encoded already and you can just call `.encode()` here. I'd leave the encoding up to Python. Alternatively, we can update `file_contents` to `Optional[Union[str, bytes]]` instead.

Comment (taljaards) [reply src/prefect/utilities/storage.py]: I went for the first option in [60c51bd](https://github.com/PrefectHQ/prefect/pull/6375/commits/60c51bd0852b764eec9f44c0b5b5f7d81b9f6be2).

**Review by madkinsz — APPROVED** [review]

**Comment by madkinsz** [comment]
@taljaards looks like sometimes `file_contents` is already `bytes`. See the failing tests. Let's just update that to a `Union` type as I mentioned in that previous comment.

**Comment by taljaards** [comment]
@madkinsz changes implemented.

**Pull request merged**